SmartDeviceLink/SDLBackgroundTaskManager.m

```diff
@@ -41,26 +41,34 @@ - (void)startBackgroundTask {
 
     __weak typeof(self) weakSelf = self;
     self.currentBackgroundTaskId = [[UIApplication sharedApplication] beginBackgroundTaskWithName:self.backgroundTaskName expirationHandler:^{
+        // We have ~1 second to do cleanup before ending the background task. If we take too long, the system will kill the app.
         __strong typeof(weakSelf) strongSelf = weakSelf;
-        SDLLogD(@"The %@ background task expired", strongSelf.backgroundTaskName);
-        [strongSelf endBackgroundTask];
+        SDLLogD(@"The background task %@ expired", strongSelf.backgroundTaskName);
+        if (strongSelf.taskEndedHandler != nil) {
+            SDLLogD(@"Waiting for cleanup to finish before ending the background task");
+            strongSelf.taskEndedHandler();
+        } else {
+            [strongSelf endBackgroundTask];
+        }
     }];
 
     SDLLogD(@"The %@ background task started with id: %lu", self.backgroundTaskName, (unsigned long)self.currentBackgroundTaskId);
 }
 
 - (void)endBackgroundTask {
+    SDLLogV(@"Attempting to end background task %@", self.backgroundTaskName);
+
     if (self.taskEndedHandler != nil) {
-        self.taskEndedHandler();
+        SDLLogD(@"Background task %@ cleanup finished", self.backgroundTaskName);
         self.taskEndedHandler = nil;
     }
```

Comment:
If this is called from the `taskEndedHandler` as I have mentioned above, then `nil`ing it out here seems invalid and the code below won't be run.

Reply:
`endBackgroundTask` does not get called from the `taskEndedHandler`. It only gets called when cleanup has finished in the `SecondaryTransportManager`'s `disconnectSecondaryTransport` method. We could add a method called `cleanupCompleted` or something like that to the background task manager.